### PR TITLE
Introduce operation aggregate nodes into graph and side panel

### PR DESCRIPTION
**Pull request by jshaughn**
This UI-side PR adds the new "Star" node to represent operation aggregate nodes.   "Operation Nodes" is a new display option that works alone or in conjunction with "Service Node" display.

Server PR: https://github.com/kiali/kiali/pull/2944
Epic #2854
Design: #2856
Fixes #2908

### CODE REVIEW

**Comment by jshaughn** [comment]
Removing DNM because checks are passing, but this is still dependent on the server PR.

**Comment by jshaughn** [comment]
@mattmahoneyrh, can start testing if you like or you may want to wait for @israel-hdez to first approve.

**Comment by mattmahoneyrh** [comment]
@jshaughn Error occured "Invalid appender [aggregateNode]" when clicking "Operation Nodes"

**Comment by jshaughn** [comment]
> @mattmahoneyrh Sounds like you don't have the server-side of the PRs.

**Comment by pbajjuri20** [comment]
Click on cars APP it needs to list all the Services, Workloads, and Operation node in the side panel.  the issue is in the side panel Operation node is not listed.   



operational node

File: src/components/CytoscapeGraph/ContextMenu/NodeContextMenu.tsx [hunk 1]
```diff
@@ -87,8 +90,8 @@ export class NodeContextMenu extends React.PureComponent<Props> {
   }
 
   render() {
-    // Disable context menu if we are dealing with a unknown or an inaccessible node
-    if (this.props.nodeType === NodeType.UNKNOWN || this.props.isInaccessible) {
+    // Disable context menu if we are dealing with an aggregate (currently has no detail) or an inaccessible node
+    if (this.props.nodeType === NodeType.AGGREGATE || this.props.isInaccessible) {
```
Comment (israel-hdez) [inline src/components/CytoscapeGraph/ContextMenu/NodeContextMenu.tsx:94]: Are `unknown` nodes inaccessible? If yes, I can understand why you removed the `this.props.nodeType === NodeType.UNKNOWN` test from the condition. If no, hmm... then I don't understand :)

**Comment by jshaughn** [comment]
> Question: why use "operation" terminology?.
"Operation" looks more the terminology used in the example of the Istio docs because they used the Open API Spec to show how to use the "request classification" feature. I think the right terminology to use is, indeed, "classification". So, I think is more right to see in the UI "classifier" = "request_operation", and "class" = "TravelQuote" as opposed to "operation" = "request_operation" and "name" = "TravelQuote".

Well this is a very good question and you make a good argument.  @jmazzitelli discussed this in the beginning and I think we were very influenced by the Istio Request Classification example.  I think the thought is that people will follow the example and adopt the "operation" terminology, as well as use "request_operation" as the introduced attribute.  If so they will expect the terminology to carry over into the UI.   I do agree that it's maybe a big assumption and that the decision is a bit arbitrary.  I think that one issue is that "Classifier Nodes" sounds a little abstract.  As you may have seen in the code, I actually use yet another term, "aggregate" to represent these nodes because in essence that is what they are at the impl level, a way to further aggregate requests.  But again, I think "Aggregate Nodes" is too abstract.  So, in the end the answer is that there is little reason to use "operation" past the Istio docs' description of the feature.  But that is perhaps more reason than I can make for any alternative.  As such I think for a first version it's reasonable and given any user adoption/feedback we can adjust going forward.

By the way, the API does provide a way to override the "request_operation" in the calls to the server.  But the initial PRs don't yet support a config option to change the default value.

> My thought: How to scale?
It's weird to see this in the graph, because you can setup multiple dimensions (so, also metrics can have several classifiers).

I really would discourage users from introducing multiple attributes to perform their request classification.  I'm not sure I see the need as the classification values could be used to provide "sub-classification" if absolutely necessary.  Again, though, this is our first support of a currently experimental feature.  We can see where it goes.

I did provide some basic graph find/hide for these nodes, from that perspective if scale.

File: src/components/CytoscapeGraph/MiniGraphCard.tsx [hunk 2]
```diff
@@ -97,6 +97,7 @@ export default class MiniGraphCard extends React.Component<MiniGraphCardProps, M
               showCircuitBreakers={false}
               showMissingSidecars={true}
               showNodeLabels={true}
+              showOperationNodes={true}
```
[the remark this answers is not in the document]
Comment (jshaughn) [reply src/components/CytoscapeGraph/MiniGraphCard.tsx:100]: I will change it to {false}, the mini-graph will stay simple and omit operation nodes for now.

File: src/components/Metrics/TrafficDetails.tsx [hunk 3]
```diff
@@ -137,12 +137,13 @@ class TrafficDetails extends React.Component<TrafficDetailsProps, TrafficDetails
 
   private buildTrafficNode = (prefix: 'in' | 'out', node: GraphNodeData): TrafficNode => {
     switch (node.nodeType) {
-      case NodeType.WORKLOAD:
+      case NodeType.AGGREGATE:
```
[the remark this answers is not in the document]
Comment (jshaughn) [reply src/components/Metrics/TrafficDetails.tsx:140]: I don't know, for now I will leave the support in the code, even if the appender is not yet used here.

File: src/pages/Graph/GraphLegendData.ts [hunk 4]
```diff
@@ -44,7 +44,7 @@ const legendData: GraphLegendItem[] = [
     data: [
       { label: 'Workload', icon: workloadImage },
       { label: 'App', icon: appImage },
-      { label: 'Unknown Source', icon: unknownSourceImage },
+      { label: 'Operation', icon: aggregateImage },
```
[the remark this answers is not in the document]
Comment (jshaughn) [reply src/pages/Graph/GraphLegendData.ts:47]: As I mentioned in the main response, I understand your point but I think (currently) there is more argument for the 'operation' nomenclature than for any alternative.

File: src/pages/Graph/GraphToolbar/GraphSettings.tsx [hunk 5]
```diff
@@ -276,6 +363,7 @@ const mapStateToProps = (state: KialiAppState) => ({
   showCircuitBreakers: state.graph.toolbarState.showCircuitBreakers,
   showMissingSidecars: state.graph.toolbarState.showMissingSidecars,
   showNodeLabels: state.graph.toolbarState.showNodeLabels,
+  showOperationNodes: state.graph.toolbarState.showOperationNodes,
```
[the remark this answers is not in the document]
Comment (jshaughn) [reply src/pages/Graph/GraphToolbar/GraphSettings.tsx:366]: This is expected and is actually an important feature point.  When service nodes are not enabled the operations represent the total traffic aggregate (across all services that may be using populating that operation name).  When service nodes are enabled you get the service-specific operation nodes, and then they can be grouped with the app.  That is why with service nodes enabled you can see the same operation show up in different app boxes.

File: src/components/CytoscapeGraph/ContextMenu/NodeContextMenu.tsx (hunk 1 above)
Comment (jshaughn) [reply src/components/CytoscapeGraph/ContextMenu/NodeContextMenu.tsx:94]: Yes, unknown nodes are always inaccessible.

**Comment by jshaughn** [comment]
@mattmahoneyrh @pbajjuri20 @israel-hdez PR review comments have been addresses.  No further changes planned unless there are further comments.

**Comment by israel-hdez** [comment]
> As such I think for a first version it's reasonable and given any user adoption/feedback we can adjust going forward.

I'm good with this. It's the "agile" way to do it :)

> I really would discourage users from introducing multiple attributes to perform their request classification. I'm not sure I see the need as the classification values could be used to provide "sub-classification" if absolutely necessary.

I wasn't thinking about sub-classes. As you say, subclasses can be achieved with just one classifier and some processing (although it could be nice tu support sub-classifiers). I was thinking more in two completely independent classes. Istio docs shows the feature with "operations", but you may also want to classify using the user kind (think about bookinfo sample where you can login as admin or regular user). In Kiali you may want to switch from viewing the "operations" or the user kind and this will let you see either which operations receive more traffic, or what are the most used services per user kind. Of course, it would also be interesting if it's possible to see both classifiers at the same time :)

But yes, let's not think about it. It's and experimental feature in Istio, so it could change.

**Pull request merged**